### DISCUSSION TITLE
Raw Packet Socket Example Socket Write Cleanup

### DISCUSSION THREAD

**Discussion by bob2oneil** [zephyrproject-rtos/zephyr#36504] [Q&A] [answered]
I am writing an application that uses a raw "packet" socket for the primary Zephyr interface.  This raw socket exposes the L2 and L3 headers of the packet.  The read portion of this is working well.  I will be using the same raw socket for both the read and write to the primary _iface_.

My question has to do with the _/zephyr/samples/net/sockets/packet_ example for the write to the socket.  In this example, 2 raw sockets are opened to the **same** _iface_, and that is a **difference** between what I intend to do, where I will use the **same** socket for both read and write.

But when a write to the socket occurs, why is the following read-back required to prevent an IP stack memory leak?

Can someone explain to me why this is true, and if it will be necessary when the read and write raw socket is the **same** singular handle?

This code sample is from the **send_packet_socket**() method:

```

	do {
		/* Sending dummy data */
		ret = sendto(packet->send_sock, lorem_ipsum, send, 0,
			     (const struct sockaddr *)&dst,
			     sizeof(struct sockaddr_ll));
		if (ret < 0) {
			LOG_ERR("Failed to send, errno %d", errno);
			break;
		} else {
			if (!FLOOD) {
				LOG_DBG("Sent %zd bytes", send);
			}
		}

		/* If we have received any data, flush it here in order to
		 * not to leak memory in IP stack.
		 */
		do {
			static char recv_buffer[RECV_BUFFER_SIZE];

			ret = recv(packet->send_sock, recv_buffer,
				   sizeof(recv_buffer), MSG_DONTWAIT);
		} while (ret > 0);


```

**Comment by jukkar** [comment, marked as answer]
It is explained in the commit 6d7f4e0b5062af360434ce5c5d4c61d8f473447f, see also some discussion on #34865

**Reply by bob2oneil** [reply]
Thanks Jukka, I will follow the recommendation and call recv() on the socket.  But assuming I am using the same socket for raw packet reception, how would I discriminate between L2 data being received by the raw packet socket, and the leak remedy?  Is it your recommendation that two separate sockets are used for the raw packet interface, per the example, a recv socket for data reception, and a send socket for transmission with the recv() fix?  This deviate a bit from my previous experiences under Linux using a raw socket, so I want to make sure I understand Zephyr best practices.  Obviously, I do not want raw packets sent to the raw socket being viewed as received packets on the same L2 interface.

**Reply by bob2oneil** [reply]
Reviewing the comments again, I fully intend to be calling recv on the same socket (if possible), but that data reception would be based presumably on incoming packets on the iface, so that activity might not be paired with sends to that same socket.  If the send and recv need to be paired to prevent the leak, then perhaps the two raw packet socket solution, one for recv, and one for tx is the solution?

**Reply by bob2oneil** [reply]
In my current testing, note that I am able to perform raw packet socket reads and writes to interfaces using a single opened socket, which is typical of Linux.
